### PR TITLE
[RFR] Fix the api

### PR DESCRIPTION
**Pull request by Luwangel**
- [x] Fix the game context
- [x] Add a cors module
- [x] Allow preflight requests

### CODE REVIEW

File: app/src/ArgumentResolver/GameContextResolver.php [hunk 1]
```diff
@@ -24,14 +26,16 @@ public function supports(Request $request, ArgumentMetadata $argument) {
 
   public function resolve(Request $request, ArgumentMetadata $argument) {
     $game = $this->gameRepository->findGameById($request->get('id'));
-    $token = $request->headers->get('Authorization');
 
-    $isPlayer = $token ? TokenAuthManager::isPlayer($request, $game, $token) : CookieAuthManager::isPlayer($request, $game);
+    $authTokenArray = explode(' ', $request->headers->get('Authorization'));
```
Comment (jdemangeon) [inline app/src/ArgumentResolver/GameContextResolver.php:30]: I don't understand this part of code easily...  A simple regex can work

File: app/src/ArgumentResolver/GameContextResolver.php [hunk 2]
```diff
@@ -24,14 +26,16 @@ public function supports(Request $request, ArgumentMetadata $argument) {
...
+    $token = count($authTokenArray) == 2 && $authTokenArray[0] == self::AUTHORIZATION_TYPE ? $authTokenArray[1] : '';
+    $isPlayer = $token != '' ? TokenAuthManager::isPlayer($request, $game, $token) : CookieAuthManager::isPlayer($request, $game);
```
Comment (jdemangeon) [inline app/src/ArgumentResolver/GameContextResolver.php]: Always use strict comp (===). Use simple comp only when necessary

File: app/src/Authentication/CookieAuthManager.php [hunk 3]
```diff
@@ -13,11 +13,17 @@ class CookieAuthManager {
...
+    $tokenCookie = $request->cookies->get(self::COOKIE_NAME);
+    $token = $tokenCookie ?: '';
+
+    return TokenAuthManager::isPlayer($request, $game, $token);
   }
 
   public static function getPlayer(Request $request, Game $game) {
-    return TokenAuthManager::getPlayer($request, $game, $request->cookies->get(self::COOKIE_NAME));
+    $tokenCookie = $request->cookies->get(self::COOKIE_NAME);
+    $token = $tokenCookie ?: '';
```
Comment (jdemangeon) [inline app/src/Authentication/CookieAuthManager.php]: Why use an empty string here ? Use null instead

File: app/src/Authentication/CookieAuthManager.php [hunk 4]
```diff
@@ -13,11 +13,17 @@ class CookieAuthManager {
   public const COOKIE_NAME = 'current-puzzle';
 
   public static function isPlayer(Request $request, Game $game) : bool {
-    return TokenAuthManager::isPlayer($request, $game, $request->cookies->get(self::COOKIE_NAME));
+    $tokenCookie = $request->cookies->get(self::COOKIE_NAME);
+    $token = $tokenCookie ?: '';
```
Comment (jdemangeon) [inline app/src/Authentication/CookieAuthManager.php]: Why use an empty string here ? Use null instead

File: app/src/Authentication/TokenAuthManager.php [hunk 5]
```diff
@@ -11,7 +11,10 @@ class TokenAuthManager {
   public static function isPlayer(Request $request, Game $game, string $token) : bool {
```
Comment (jdemangeon) [inline app/src/Authentication/TokenAuthManager.php]: string $token = null

File: app/src/Controller/ApiController.php [hunk 6]
```diff
@@ -31,7 +31,12 @@ public function __construct(GameApi $gameApi, GameRepository $gameRepository, En
   }
 
   public function new(Request $request) {
-    $apiResponse = $this->api->new(self::DEFAULT_SIZE, $request->get('mode') === 'multi');
+    if ($request->getMethod() == 'OPTIONS') {
```
Comment (jdemangeon) [inline app/src/Controller/ApiController.php]: I've never seen this kind of code in controller before... Your controller need to be as simple as possible. It must include the minimum logic as possible.

File: app/src/Controller/ApiController.php [hunk 7]
```diff
@@ -42,7 +47,11 @@ public function new(Request $request) {
     ]);
   }
 
-  public function game(GameContext $context) {
+  public function game(Request $request, GameContext $context) {
+    if ($request->getMethod() == 'OPTIONS') {
```
Comment (jdemangeon) [inline app/src/Controller/ApiController.php]: Same, you don't need to deal with OPTIONS....

File: app/src/Controller/ApiController.php [hunk 8]
```diff
@@ -56,44 +65,46 @@ public function game(GameContext $context) {
...
       'id' => $game->getId(),
+      'isMultiplayer' => $game->getIsMultiplayer(),
       'currentPlayer' => $player,
       'winner' => $game->getWinner()
     ]);
   }
 
-  public function cancel(GameContext $context) {
+  public function cancel(Request $request, GameContext $context) {
+    if ($request->getMethod() == 'OPTIONS') {
```
Comment (jdemangeon) [inline app/src/Controller/ApiController.php]: Same

File: app/src/Controller/ApiController.php [hunk 9]
```diff
@@ -56,44 +65,46 @@ public function game(GameContext $context) {
...
+
     if ($context->getIsPlayer()) {
       $this->em->remove($context->getGame());
       $this->em->flush();
 
       return new Response(
-        'Content',
-        Response::HTTP_OK,
-        ['content-type' => 'text/html']
+        Response::HTTP_OK
```
Comment (jdemangeon) [inline app/src/Controller/ApiController.php]: The first arg is the content (the second is the status code), you send "200" in the body here

File: app/src/Controller/ApiController.php [hunk 10]
```diff
@@ -56,44 +65,46 @@ public function game(GameContext $context) {
...
       );
     }
-    return new Response(
-      'Content',
-      Response::HTTP_INTERNAL_SERVER_ERROR,
-      ['content-type' => 'text/html']
-    );
+    return new Response(Response::HTTP_INTERNAL_SERVER_ERROR);
```
Comment (jdemangeon) [inline app/src/Controller/ApiController.php]: Same with "500"

File: app/src/Controller/ApiController.php [hunk 11]
```diff
@@ -56,44 +65,46 @@ public function game(GameContext $context) {
...
   }
 
-  public function join(GameContext $context, TokenGenerator $tokenGenerator) {
+  public function join(Request $request, GameContext $context, TokenGenerator $tokenGenerator) {
+    if ($request->getMethod() == 'OPTIONS') {
```
Comment (jdemangeon) [inline app/src/Controller/ApiController.php]: Same for responsabilities

File: app/src/Controller/ApiController.php [hunk 12]
```diff
@@ -56,44 +65,46 @@ public function game(GameContext $context) {
...
+    }
+
     $game = $context->getGame();
 
     if ($game->isFull()) {
       return new Response(
-        'Content',
-        Response::HTTP_INTERNAL_SERVER_ERROR,
-        ['content-type' => 'text/html']
+        Response::HTTP_INTERNAL_SERVER_ERROR
```
Comment (jdemangeon) [inline app/src/Controller/ApiController.php]: First arg is the content

File: app/src/Controller/ApiController.php [hunk 13]
```diff
@@ -113,29 +124,37 @@ public function join(GameContext $context, TokenGenerator $tokenGenerator) {
     ]);
   }
 
-  public function move(GameContext $context, int $tile) {
+  public function move(Request $request, GameContext $context, int $tile) {
+    if ($request->getMethod() == 'OPTIONS') {
```
Comment (jdemangeon) [inline app/src/Controller/ApiController.php]: Same for responsabilities

File: app/src/Controller/ApiController.php [hunk 14]
```diff
@@ -113,29 +124,37 @@ public function join(GameContext $context, TokenGenerator $tokenGenerator) {
...
-      'otherPlayer' => $this->serializer->serialize($otherPlayer, 'json'),
-      'winner' => $this->serializer->serialize($game->getWinner(), 'json')
+      'currentPlayer' => $currentPlayer,
+      'winner' => $game->getWinner()
     ]);
   }
 
-  public function games() {
+  public function games(Request $request) {
+    if ($request->getMethod() == 'OPTIONS') {
```
Comment (jdemangeon) [inline app/src/Controller/ApiController.php]: Same for responsabilities

File: app/src/Security/CorsListener.php [hunk 15]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace App\Security;
+
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
+
+class CorsListener {
+    public function onKernelResponse(FilterResponseEvent $event) {
+        $responseHeaders = $event->getResponse()->headers;
```
Comment (jdemangeon) [inline app/src/Security/CorsListener.php]: :+1:  Good, listeners let us apply global logic to our requests

**Review by fzaninotto — APPROVED** [review]

**Pull request merged**